R/arc-select.R

```diff
@@ -18,50 +20,20 @@ arc_select <- function(
     filter_geom,
     predicate = "intersects",
     n_max = Inf,
-    ...
+    ...,
+    error_call = rlang::caller_env()
 ) {
 
-  stopifnot(inherits(x, c("Table", "FeatureLayer")))
+  obj_check_arcgis(x, call = error_call)
```

Comment:
If the only place where these checks are happening is inside of `arc_select()`, I do not think it fully justifies creating an entirely new helper function. If there is another place where we may want to provide a filter geometry, then the helper ought to be added to arcgisutils.

Reply:
I figured you'd want flexibility to do checks against all the different types in the future. Here is a list of options I gleaned from the REST API docs sometime last year: "MapServer", "FeatureServer", "GPServer", "GeocodeServer", "GeometryServer", "ImageServer". I can definitely move this to utils.R but I'd like to keep check in the name if it is going to trigger an error. The naming convention of `obj_check` is based on similar packages in `{vctrs}` but I could just call it `check_arcgis_service()`. Although I'd note that I'm also using it to check the class of input objects so maybe something more generic like `check_inherits_any()` could work.

Reply:
Good point about the prefix `obj_check`. It isn't clear to me how this funciton can be generalized. It feels like it is a flexible wrapper around `stopifnot(inherits(x, class))` but using cli. For example the use case in `apply_filter_geom` uses the classes `c("sfc", "sfg", "bbox")` which is not any of the classes listed above and feels out of scope of the intent. 

In this very specific example, we want to error if it is not a `Table` or  `FeatureLayer`. If it is of class `ImageServer` we _do_ want to error. When we look at `vctrs::obj_check_list()` there is a clear type we are checking against. One possibility would be to provide utilities like `obj_check_layer()` and `obj_check_server()` to check for `Table`, `FeatureLayer` and `FeatureServer`, `MapServer`, and `ImageServer` respectively. 

How can we be _very_ explicit about the use cases here?

Reply:
Yeah – I typically start with `stopifnot()` calls but switch my input checks use `cli_abort()` to take advantage of the `{cli}` message formatting.

The idea of `obj_check_layer()` and `obj_check_server()` seems like a good approach. If you're checking for sfc or sfg input classes in any other functions, I think a helper function would be useful there as well but I'm fine if you'd rather limit the number of tiny functions you need to keep track of.